pkg/adapter/containers.go

```diff
@@ -1288,3 +1297,306 @@ func (r *LocalRuntime) Commit(ctx context.Context, c *cliconfig.CommitValues, co
+	hostUID, hostGID, err := utils.GetHostIDs(util.ConvertIDMap(idMappingOpts.UIDMap), util.ConvertIDMap(idMappingOpts.GIDMap), user.UID, user.GID)
+	if err != nil {
+		return err
+	}
+
+	hostOwner := idtools.IDPair{UID: int(hostUID), GID: int(hostGID)}
+
+	if isFromHostToCtr {
+		if isVol, volDestName, volName := isVolumeDestName(destPath, ctr); isVol {
+			path, err := pathWithVolumeMount(ctr, r.Runtime, volDestName, volName, destPath)
+			if err != nil {
+				return errors.Wrapf(err, "error getting destination path from volume %s", volDestName)
+			}
+			destPath = path
+		} else if isBindMount, mount := isBindMountDestName(destPath, ctr); isBindMount {
+			path, err := pathWithBindMountSource(mount, destPath)
+			if err != nil {
+				return errors.Wrapf(err, "error getting destination path from bind mount %s", mount.Destination)
+			}
+			destPath = path
+		} else if filepath.IsAbs(destPath) {
+			cleanedPath, err := securejoin.SecureJoin(mountPoint, destPath)
+			if err != nil {
+				return err
+			}
+			destPath = cleanedPath
+
+		} else {
+			ctrWorkDir, err := securejoin.SecureJoin(mountPoint, ctr.WorkingDir())
+			if err != nil {
+				return err
+			}
+			if err = idtools.MkdirAllAndChownNew(ctrWorkDir, 0755, hostOwner); err != nil {
+				return errors.Wrapf(err, "error creating directory %q", destPath)
+			}
+			cleanedPath, err := securejoin.SecureJoin(mountPoint, filepath.Join(ctr.WorkingDir(), destPath))
+			if err != nil {
+				return err
+			}
+			destPath = cleanedPath
+		}
+	} else {
+		destOwner = idtools.IDPair{UID: os.Getuid(), GID: os.Getgid()}
+		var err error
+		if isVol, volDestName, volName := isVolumeDestName(srcPath, ctr); isVol {
+			srcPath, err = pathWithVolumeMount(ctr, r.Runtime, volDestName, volName, srcPath)
+			if err != nil {
+				return errors.Wrapf(err, "error getting source path from volume %s", volDestName)
+			}
+		} else if isBindMount, mount := isBindMountDestName(srcPath, ctr); isBindMount {
+			srcPath, err = pathWithBindMountSource(mount, srcPath)
+			if err != nil {
+				return errors.Wrapf(err, "error getting source path from bind mount %s", mount.Destination)
+			}
+		} else if filepath.IsAbs(srcPath) {
+			srcPath, err = securejoin.SecureJoin(mountPoint, srcPath)
+			if err != nil {
+				return err
+			}
+		} else {
+			srcPath, err = securejoin.SecureJoin(mountPoint, filepath.Join(ctr.WorkingDir(), srcPath))
+			if err != nil {
+				return err
+			}
+		}
+	}
+	if !filepath.IsAbs(destPath) {
+		dir, err := os.Getwd()
+		if err != nil {
+			return errors.Wrapf(err, "err getting current working directory")
+		}
+		destPath = filepath.Join(dir, destPath)
+	}
+
+	if src == "-" {
+		srcPath = os.Stdin.Name()
+		extract = true
+	}
+	return r.Copy(srcPath, destPath, dest, idMappingOpts, &destOwner, extract, isFromHostToCtr)
+}
+
+// ParsePath parses the input of cp command returns the path and the container
+func (r *LocalRuntime) ParsePath(path string) (*Container, string) {
```

Comment:
If you move this code to pkg/util/utils.go you won't need to duplicate it between local and remote.

Reply:
there will be an import circle cmd/podman imports adapter, adapter imports util, util imports adapter. Because ParsePath needs adapter.Runtime to check the input container. Can I get rid of this circle?